Index creation waits for write consistency shards

abeyad opened the pull request:
Before returning, index creation now waits for the write consistency
number of shards to be available. An index can not take any indexing or
other replication operations without the write consistency level of
shards being available anyway, so waiting on the index creation response
in order for this condition to be met makes sense, and allows API users
to not depend on cluster health checks before attempting indexing
operations on the newly created index.

nik9000 commented on core/src/main/java/org/elasticsearch/action/admin/indices/create/CreateIndexRequest.java:
```diff
@@ -77,6 +79,8 @@
 
     private boolean updateAllTypes = false;
 
+    private WriteConsistencyLevel waitOnShards = WriteConsistencyLevel.DEFAULT;
```
This seems like an odd name. Maybe call it `waitForConsistency`?

nik9000 commented on core/src/main/java/org/elasticsearch/action/admin/indices/create/CreateIndexRequest.java:
```diff
@@ -440,6 +444,21 @@ public CreateIndexRequest updateAllTypes(boolean updateAllTypes) {
+     * {@link ReplicationOperation#WRITE_CONSISTENCY_LEVEL_SETTING} setting.
+     */
+    public CreateIndexRequest waitOnShards(String waitOnShards) {
+        this.waitOnShards = WriteConsistencyLevel.fromString(waitOnShards);
```
Why not take a `WriteConsistencyLevel`?

nik9000 commented on core/src/main/java/org/elasticsearch/action/admin/indices/shrink/ShrinkRequest.java:
```diff
@@ -138,4 +138,14 @@ public void source(BytesReference source) {
+     * {@link ReplicationOperation#WRITE_CONSISTENCY_LEVEL_SETTING} setting.
+     */
+    public void setWaitOnShards(String waitOnShards) {
+        this.shrinkIndexRequest.waitOnShards(Objects.requireNonNull(waitOnShards));
```
I think if you are going to go to the trouble to call `requireNonNull` you should have an error message. Either we expect it to throw and NPE in the parser or stick a message on the exception.

nik9000 commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexCreationMonitor.java:
```diff
@@ -0,0 +1,132 @@
+
+        if (observedState.status() == ClusterState.ClusterStateStatus.APPLIED
+                && enoughShardsActive(observedState, indexName, writeConsistencyLevel)) {
+            observerListener.onNewClusterState(observedState);
```
Maybe this check should be above the listener/predicate construction? Sort of a "quick return" style?

nik9000 commented on core/src/test/java/org/elasticsearch/action/admin/indices/create/CreateIndexIT.java:
```diff
@@ -477,4 +476,22 @@ public void testCreateShrinkIndexFails() throws Exception {
+        for (int i = 0; i < numDocs; i++) {
+            client().prepareIndex(indexName, "type").setSource("{\"foo\" : \"bar\", \"i\" : " + i + "}").get();
+        }
+        assertThat(client().prepareSearch("target").setSize(numDocs).get().getHits().getTotalHits(), equalTo(numDocs));
```
I think there is an `assertHitCount` you can use. Do you want `setSize(0)` as well?

nik9000 commented:
Wonderful news! I don't know a ton about the cluster state stuff but it seems sane and after applying it this gist passes:
https://gist.github.com/nik9000/3d65d5d883c7fd918a0835947c6549eb

ywelsch commented on core/src/main/java/org/elasticsearch/action/admin/indices/create/TransportCreateIndexAction.java:
```diff
@@ -83,15 +94,34 @@ protected void masterOperation(final CreateIndexRequest request, final ClusterSt
             @Override
             public void onResponse(ClusterStateUpdateResponse response) {
-                listener.onResponse(new CreateIndexResponse(response.isAcknowledged()));
+                // the cluster state that includes the newly created index has been acknowledged,
```
I wonder if we should put this into MetaDataCreateIndexService instead (so we don't need to duplicate it into the two transport actions).

abeyad replied:
@ywelsch I actually did this in a new commit which I will push up soon.

ywelsch commented on core/src/main/java/org/elasticsearch/action/WriteConsistencyLevel.java:
```diff
@@ -67,4 +73,12 @@ public static WriteConsistencyLevel fromString(String value) {
         throw new IllegalArgumentException("No write consistency match [" + value + "]");
     }
+
+    public void writeTo(StreamOutput out) throws IOException {
```
add `@Override` annotation

ywelsch commented on core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java:
```diff
@@ -50,6 +51,12 @@
             PrimaryResultT extends ReplicationOperation.PrimaryResult<ReplicaRequest>
         > {
+
+    public static final Setting<WriteConsistencyLevel> WRITE_CONSISTENCY_LEVEL_SETTING =
```
this setting is not registered (and thus not working in ES 5.0)

ywelsch commented on core/src/main/java/org/elasticsearch/action/admin/indices/create/CreateIndexRequest.java:
```diff
@@ -440,6 +443,28 @@ public CreateIndexRequest updateAllTypes(boolean updateAllTypes) {
         return this;
     }
 
+    public int waitOnActiveShards() {
```
There is a superclass ReplicationRequest where we should put these methods (as they apply to all the replication requests).

abeyad replied:
`CreateIndexRequest` does not extend `ReplicationRequest`, nor does `UpdateRequest`, etc.

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java:
```diff
@@ -176,7 +178,12 @@ public void validateIndexName(String index, ClusterState state) {
-    public void createIndex(final CreateIndexClusterStateUpdateRequest request, final ActionListener<ClusterStateUpdateResponse> listener) {
+    public void createIndex(final CreateIndexClusterStateUpdateRequest request,
+                            final ActionListener<ClusterStateUpdateResponse> listener,
+                            final ThreadContext threadContext) {
```
I think we should add ThreadPool as dependency to MetaDataCreateIndexService instead.

ywelsch commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -89,7 +90,7 @@
     final protected IndicesService indicesService;
     final private ShardStateAction shardStateAction;
-    final private WriteConsistencyLevel defaultWriteConsistencyLevel;
+    final private int defaultWaitOnActiveShards;
```
I think we agreed on making this setting updatable (cluster setting).

bleskes replied:
indeed, but I think it’s good to do it in a different change?

> On 24 Jun 2016, at 10:08, Yannick Welsch <notifications@github.com> wrote:
> 
> In core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
> 
> > @@ -89,7 +90,7 @@
> >      final protected ClusterService clusterService;
> >      final protected IndicesService indicesService;
> >      final private ShardStateAction shardStateAction;
> > -    final private WriteConsistencyLevel defaultWriteConsistencyLevel;
> > +    final private int defaultWaitOnActiveShards;
> 
> I think we agreed on making this setting updatable (cluster setting).
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

ywelsch commented on core/src/main/java/org/elasticsearch/action/update/UpdateRequestBuilder.java:
```diff
@@ -122,10 +121,16 @@ public UpdateRequestBuilder setVersionType(VersionType versionType) {
      */
-    public UpdateRequestBuilder setConsistencyLevel(WriteConsistencyLevel consistencyLevel) {
-        request.consistencyLevel(consistencyLevel);
+    public UpdateRequestBuilder setWaitOnActiveShards(int waitOnActiveShards) {
```
We also have `ReplicationRequestBuilder`. No need to add this to individual builders.

ywelsch commented on core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java:
```diff
@@ -50,6 +50,11 @@
         > {
+
+    public static final int WAIT_ON_ACTIVE_SHARDS_NOT_SET = 0;
+    public static final Setting<Integer> WAIT_ON_ACTIVE_SHARDS_SETTING =
```
I wonder if we should make this a String setting that allows values such as `all`.

ywelsch commented on core/src/main/java/org/elasticsearch/action/admin/indices/create/CreateIndexRequest.java:
```diff
@@ -440,6 +443,28 @@ public CreateIndexRequest updateAllTypes(boolean updateAllTypes) {
+     * @param waitOnActiveShards number of active shard copies to wait on, must be between 0 and numShardCopies, or -1 for all.
+     */
+    public CreateIndexRequest waitOnActiveShards(int waitOnActiveShards) {
+        if (waitOnActiveShards == 0 || waitOnActiveShards < -1) {
```
Why is it not possible to specify 0? I might want to create an index without waiting for any shard of that index to be active.

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexCreationMonitor.java:
```diff
@@ -0,0 +1,131 @@
+    private final ClusterService clusterService;
+    private final ThreadContext threadContext;
+
+    protected IndexCreationMonitor(Settings settings, ClusterService clusterService, ThreadContext threadContext) {
```
We should use the thread context of the caller of `waitOnShards`, not the one passed here when we construct `IndexCreationMonitor`. Make this `ThreadPool` instead.

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexCreationMonitor.java:
```diff
@@ -0,0 +1,131 @@
+
+            @Override
+            public void onClusterServiceClose() {
+                logger.warn("[{}] cluster service closed while waiting for write consistency number of shards " +
```
no need to log warnings when we shut down node. `debug` logging is good enough.

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexCreationMonitor.java:
```diff
@@ -0,0 +1,131 @@
+            public void onClusterServiceClose() {
+                logger.warn("[{}] cluster service closed while waiting for write consistency number of shards " +
+                                "to be allocated on index creation.", indexName);
+                listener.onResponse(new ClusterStateUpdateResponse(false));
```
We should notify the client about the node closing instead: `listener.onFailure(new NodeClosedException(clusterService.localNode()))`

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexCreationMonitor.java:
```diff
@@ -0,0 +1,131 @@
+            public void onTimeout(TimeValue timeout) {
+                logger.warn("[{}] index was created, but the operation timed out waiting for write consistency " +
+                                "number of shards to be allocated.", indexName);
+                listener.onResponse(new ClusterStateUpdateResponse(false));
```
I would also use `listener.onFailure` here...

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexCreationMonitor.java:
```diff
@@ -0,0 +1,131 @@
+
+        if (observedState.status() == ClusterState.ClusterStateStatus.APPLIED
+                && enoughShardsActive(observedState, indexName, settings, waitOnActiveShards)) {
+            observerListener.onNewClusterState(observedState);
```
I wonder if repeating the Change predicate here is needed. When we call observer.waitForNextChange() and the predicate holds, we directly call `onNewClusterState` on the listener...

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java:
```diff
@@ -190,6 +197,27 @@ protected ClusterStateUpdateResponse newResponse(boolean acknowledged) {
+                        } else {
+                            final TimeValue remainingTime = TimeValue.timeValueNanos(timeToTimeout);
+                            try {
+                                IndexCreationMonitor monitor = new IndexCreationMonitor(settings, clusterService, threadContext);
```
ok, I see now that we create one monitor per index creation. Is that necessary? We could just have a single IndexCreationMonitor instance as the class is pretty stateless.

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexCreationMonitor.java:
```diff
@@ -0,0 +1,131 @@
+        }
+        final IndexRoutingTable indexRoutingTable = clusterState.routingTable().index(indexName);
+        assert indexRoutingTable != null;
+        if (indexRoutingTable.allPrimaryShardsActive() == false) {
```
what if I don't want to require for any shards to be available after index creation?

ywelsch commented on core/src/test/java/org/elasticsearch/action/admin/indices/create/CreateIndexWaitsIT.java:
```diff
@@ -0,0 +1,155 @@
+
+/**
+ * Tests that the index creation operation waits for the appropriate
+ * consistency level of shards to be started before returning, or
```
remove "consistency level"

ywelsch commented on core/src/test/java/org/elasticsearch/action/admin/indices/create/CreateIndexWaitsIT.java:
```diff
@@ -0,0 +1,155 @@
+ * consistency level of shards to be started before returning, or
+ * times out as appropriate.
+ */
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST, numDataNodes = 1)
```
any reason that this test is restricted to a single data node?

abeyad replied:
That is just to start off with for each test... then the individual tests add nodes as needed.  As far as I can tell, that's how `InternalTestCluster` works - it uses `numDataNodes` in the annotation as the initial number of nodes to create the cluster with.  

I've also added asserts, to ensure we have the appropriate number of data nodes, and those asserts pass.

ywelsch replied:
Is there a reason to start off with 1? If you don't specify it, it's up to the cluster to pick the number of nodes randomly.

ywelsch commented on core/src/test/java/org/elasticsearch/action/admin/indices/create/CreateIndexWaitsIT.java:
```diff
@@ -0,0 +1,155 @@
+                                .put(INDEX_NUMBER_OF_SHARDS_SETTING.getKey(), randomIntBetween(2, 5))
+                                .put(INDEX_NUMBER_OF_REPLICAS_SETTING.getKey(), numReplicas)
+                                .build();
+        internalCluster().ensureAtMostNumDataNodes(numReplicas);
```
will this do anything? The test class specifies that we have exactly 1 data node

ywelsch commented on core/src/test/java/org/elasticsearch/action/admin/indices/create/CreateIndexWaitsIT.java:
```diff
@@ -0,0 +1,155 @@
+                      .setWaitOnActiveShards(-1)
+                      .setTimeout("2s");
+        assertTrue(request.get().isAcknowledged());
+        deleteIndex(indexName);
```
Tests will automatically clean up indices.

ywelsch commented on core/src/test/java/org/elasticsearch/action/admin/indices/create/CreateIndexWaitsIT.java:
```diff
@@ -0,0 +1,155 @@
+                                .build();
+        CreateIndexRequestBuilder request = prepareCreate(indexName).setSettings(settings).setWaitOnActiveShards(-2);
+        assertFalse(request.get().isAcknowledged());
+        deleteIndex(indexName);
```
we should check here that no index was created?

ywelsch commented on core/src/test/java/org/elasticsearch/cluster/allocation/ClusterRerouteIT.java:
```diff
@@ -253,14 +253,14 @@ private void rerouteWithAllocateLocalGateway(Settings commonSettings) throws Exc
+            String nodeId = clusterState.nodes().resolveNode(nodeToCheck).getId();
+            return clusterState.getRoutingNodes().node(nodeId).iterator().next().state() == ShardRoutingState.STARTED;
+        });
+        assertTrue(result);
```
use assertBusy instead.

ywelsch commented on core/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java:
```diff
@@ -144,6 +145,7 @@ protected void doExecute(Task task, BulkRequest bulkRequest, ActionListener<Bulk
                     }
                     createIndexRequest.cause("auto(bulk api)");
                     createIndexRequest.masterNodeTimeout(bulkRequest.timeout());
+                    createIndexRequest.waitForActiveShards(ActiveShardCount.NONE);
```
What's the reason not to use `bulkRequest.waitForActiveShards()` here?

abeyad replied:
Because the bulk request would have to wait for the same amount of shards anyway during write (it will retry if the shards aren't available)

ywelsch commented on core/src/main/java/org/elasticsearch/action/index/TransportIndexAction.java:
```diff
@@ -94,6 +95,7 @@ protected void doExecute(Task task, final IndexRequest request, final ActionList
             createIndexRequest.mapping(request.type());
             createIndexRequest.cause("auto(index api)");
             createIndexRequest.masterNodeTimeout(request.timeout());
+            createIndexRequest.waitForActiveShards(ActiveShardCount.NONE);
```
same question here, why not `request.waitForActiveShards()`?

ywelsch commented on core/src/main/java/org/elasticsearch/action/support/ActiveShardCount.java:
```diff
@@ -0,0 +1,181 @@
+    public int resolve(final IndexShardRoutingTable shardRoutingTable) {
+        Objects.requireNonNull(shardRoutingTable);
+        if (this == ActiveShardCount.DEFAULT) {
+            int val = shardRoutingTable.size() / 2; // default is half the shards must be active
```
Should this be `(shardRoutingTable.size() + 1) / 2` with special case for `shardRoutingTable.size == 2`?

ywelsch replied:
I wonder if we should just let the default be 1 instead of confusing the user with some sort of QUORUM semantics that does not exist in our replication model.

abeyad replied:
> Should this be (shardRoutingTable.size() + 1) / 2 with special case for shardRoutingTable.size == 2?

I think its debatable.  If `shardRoutingTable.size()` is 5, then we will return 3 instead of 2, which is more "quorum" than "half".  I don't mind if the semantics are "at least half".

> I wonder if we should just let the default be 1 instead of confusing the user with some sort of QUORUM semantics that does not exist in our replication model.

@bleskes suggested the default be "half" instead of 1.  I think if we view it as a "don't let a write proceed if we are down to just 1 shard", then maybe "half" is safer as a default, without introducing the word "quorum" or a notion of quorum, which may be confusing to the user.

bleskes replied:
> just let the default be 1 instead 

My rational with going with half as default is that I think that adding replicas should change the behavior - if someone runs with 6 copies , it's probably not a good default to let of them (but one) go away before signalling alarm.

I chose the word "half" in order to avoid a loaded word like "quorum" which implies stuff that aren't part of our model (i.e., quorum reads). I don't mind if we round up (i.e., `(size() + 1) / 2`) or down (i.e. `size()/2` ) as long as it's not `size()/2 + 1` .

ywelsch replied:
6 shard copies? That's rather useless in a system like ES. Reasonable choices are 1, 2 or 3 shard copies.

For 1 copy, it trivially makes sense to choose wait_for_active_shards = 1 as default. Fixing defaults for 2 or 3 shard copies feels highly arbitrary to me, there is no good default for all situations. If we always pick wait_for_active_shards = 1 as default, it is easy to document, simple to understand and does not imply any kind of quorum semantics.

This setting is of limited use anyhow as it does not provide the guarantee that most users are after, namely that a write has made it to a certain number of shard copies once we acknowledge the write.  Maybe the better approach would be to remove this setting for now (at the replication operation level, not index creation) until we've implemented one that gives guarantees on the number of shard copies that have received a write.

bleskes replied:
> 6 shard copies? That's rather useless in a system like ES

We do have people using more then 3 shards so that lead to the idea of having the default scale with it. Thinking about it more I think the main usage for having so many copies is auto-expand-replicas-like usages, where you want to have shard copies available on all active nodes. In those case I think you mostly care about the data being on everything that’s up and not be bound by durability guarantees. In that case I would be fine with waitForActiveShards default to 1 and allow people to set things differently on the index level if the want different default (when we do that change).

> This setting is of limited use anyhow as it does not provide the guarantee that most users are after

Correct - this setting is meant to be used to limit the scope of events that will be indexed into less than a given number of shards. It should be coupled with a check of the response of each write operation.

abeyad replied:
>  In that case I would be fine with waitForActiveShards default to 1 

It makes sense to me as well, given the simplicity of the semantics of just using 1 as the default (less complicated to explain/understand for users).

ywelsch commented on core/src/main/java/org/elasticsearch/action/support/ActiveShardCount.java:
```diff
@@ -0,0 +1,181 @@
+     * The default value indicates half the shards must be active, but if there is only one shard
+     * in the index, then the value for the default will be 1 (instead of 0).
+     */
+    public int resolve(final IndexShardRoutingTable shardRoutingTable) {
```
This method could take an IndexMetaData object as parameter instead. This would let us get rid of exceeds method as well.

abeyad replied:
Good point, will do

ywelsch commented on core/src/main/java/org/elasticsearch/action/support/ActiveShardCount.java:
```diff
@@ -0,0 +1,181 @@
+            try {
+                val = Integer.parseInt(str);
+            } catch (NumberFormatException e) {
+                throw new IllegalArgumentException("Cannot parse ActiveShardCount[" + str + "]");
```
exception messages should start with lowercase (for consistency)

ywelsch commented on core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java:
```diff
@@ -95,7 +104,7 @@ public ReplicationOperation(Request request, Primary<Request, ReplicaRequest, Pr
 
     public void execute() throws Exception {
-        final String writeConsistencyFailure = checkWriteConsistency ? checkWriteConsistency() : null;
+        final String writeConsistencyFailure = checkWaitForActiveShards ? checkWaitForActiveShards() : null;
```
rename this variable (`writeConsistencyFailure`)?

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java:
```diff
@@ -176,7 +192,17 @@ public void validateIndexName(String index, ClusterState state) {
+                            final ActionListener<ClusterStateUpdateResponse> listener) {
+
+        final int numShardCopies = IndexMetaData.INDEX_NUMBER_OF_REPLICAS_SETTING.get(request.settings()) + 1;
+        if (request.waitForActiveShards().exceeds(numShardCopies)) {
```
I think this check comes too early. Templates have not been applied yet. I suggest doing this once IndexMetadata has been created.

bleskes replied:
Maybe it's the simplest to add the waiting to the TransportCreateIndexAction ? is there any good reason to inline it here?

abeyad replied:
@bleskes There would be some tradeoffs in moving the waiting to the `TransportCreateIndexAction`.  It will indeed simplify the code, but there are other actions other than `TransportCreateIndexAction` that call `MetaDataCreateIndexService#createIndex`, namely `TransportRolloverAction` and `TransportShrinkAction`.  It may be correct that neither of those two actions should do any waiting anyway.  But I was also thinking, now that our default wait_for_active_shards is 1, why not have all transport actions (TransportBulkAction, TransportUpdateAction, etc) just wait using the default (instead of specifying NONE) as it would simplify things?

The other tradeoff is, if the user specifies wait_for_active_shards > number_of_replicas + 1, here we can throw an error on index creation itself with a message that the request itself has something wrong in it.  If we move the waiting to the TransportCreateIndexAction, then we can only validate the wait_for_active_shards part of the request after the index has been created.

abeyad replied:
Actually, I'm able to easily keep the check while separating the waiting part out into the `TransportCreateIndexAction`.  It does make the code simpler and less overall changes.  The question is, should `TransportRolloverAction` wait for active shards just like `TransportCreateIndexAction`?

bleskes replied:
I think it makes sense that the Rollover API will wait. The Shrink API should also wait IMO, but sometimes it may copy lots of files (if hard links are not supported by the FS).. I'm inclined to say both should wait and if we see that the Shrink API times out more often than not, we can revisit.  I didn't read the PR here to voice a proper opinion whether it's simpler to have these 3 actions wait independently (using the same share "waiting" class/cluster state predicate) or have it uniform here (with the risk of people waiting where the shouldn't). My feeling says wait in the action - but that's for your consideration.

ywelsch commented on core/src/main/java/org/elasticsearch/action/support/replication/ReplicationRequest.java:
```diff
@@ -131,11 +131,18 @@ ShardId shardId() {
-    @SuppressWarnings("unchecked")
-    public final Request consistencyLevel(WriteConsistencyLevel consistencyLevel) {
-        this.consistencyLevel = consistencyLevel;
+    public final Request waitForActiveShards(ActiveShardCount waitForActiveShards) {
```
add back `@SuppressWarnings("unchecked")`

ywelsch commented on core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:
```diff
@@ -89,7 +91,7 @@
     final protected IndicesService indicesService;
     final private ShardStateAction shardStateAction;
-    final private WriteConsistencyLevel defaultWriteConsistencyLevel;
+    final private ActiveShardCount defaultWaitForActiveShards;
```
I though that we wanted this setting to be dynamically adjustable?

abeyad replied:
we do, i was going to put it in a separate commit, i'll add it soon.

bleskes replied:
+1 to a separate commit

ywelsch commented on core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java:
```diff
@@ -50,6 +51,14 @@
+                      (String) null,
+                      ActiveShardCount::parseString,
+                      Setting.Property.Dynamic,
+                      Setting.Property.NodeScope);
```
We should think if we want to make this an index-level setting instead (It is often directly related to the number of replicas that an index has).

bleskes replied:
+1 to index level. If possible, and unless it simplifies something here,  let's do it as a follow up - this PR is getting big :)

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java:
```diff
@@ -190,6 +216,28 @@ protected ClusterStateUpdateResponse newResponse(boolean acknowledged) {
                     @Override
+                    public void onAllNodesAcked(@Nullable Throwable t) {
+                        // the cluster state that includes the newly created index has been acknowledged,
+                        // now wait for the write consistency number of shards to be allocated before returning,
```
"write consistency" ?

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java:
```diff
@@ -190,6 +216,28 @@ protected ClusterStateUpdateResponse newResponse(boolean acknowledged) {
+                        // the cluster state that includes the newly created index has been acknowledged,
+                        // now wait for the write consistency number of shards to be allocated before returning,
+                        // as that is when indexing operations can take place on the newly created index
+                        final long timeToTimeout = request.ackTimeout().nanos() - (System.nanoTime() - startTimeNS);
```
The time span sampled here is from before we submit cluster state update until acking completed. There are two timeouts however that come into play when submitting cluster state updates, ackTimeout and masterNodeTimeout. I would make things simpler here and proceed in the same way that we do in other places: just do waitOnShards() with masterNodeTimeout.

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java:
```diff
@@ -550,4 +598,72 @@ static void prepareShrinkIndexSettings(ClusterState currentState, Set<String> ma
+                              final ActiveShardCount waitForActiveShards,
+                              final ActionListener<ClusterStateUpdateResponse> listener,
+                              final TimeValue timeout,
+                              final ThreadContext threadContext) {
```
no need to pass ThreadContext here. Just do `threadPool.getThreadContext()` when you create the observer.

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java:
```diff
@@ -550,4 +598,72 @@ static void prepareShrinkIndexSettings(ClusterState currentState, Set<String> ma
+            @Override
+            public boolean apply(ClusterState previousState, ClusterState.ClusterStateStatus previousStatus,
+                                 ClusterState newState, ClusterState.ClusterStateStatus newStatus) {
+                return newStatus == ClusterState.ClusterStateStatus.APPLIED
```
why does it need to be applied?

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexCreationMonitor.java:
```diff
@@ -86,32 +86,30 @@ public void onNewClusterState(ClusterState state) {
+            waitForActiveShards = WAIT_FOR_ACTIVE_SHARDS_SETTING.get(settings);
         }
         final IndexRoutingTable indexRoutingTable = clusterState.routingTable().index(indexName);
         assert indexRoutingTable != null;
```
The `onAllNodesAcked` can be called before we've actually set the cluster state on this node (this should happen very rarely though). This means that the newly created index is not part yet of the cluster state when we come here.

ywelsch replied:
what I wrote here is not true. The master node itself only acks once it has applied the state locally.

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/metadata/IndexCreationMonitor.java:
```diff
@@ -86,32 +86,30 @@ public void onNewClusterState(ClusterState state) {
     }
 
     // package private for testing
     static boolean enoughShardsActive(final ClusterState clusterState, final String indexName,
```
I wonder if we should move this method (and checkWaitForActiveShards in ReplicationOperation) directly into ActiveShardCount

ywelsch commented on core/src/test/java/org/elasticsearch/action/admin/indices/create/CreateIndexWaitsIT.java:
```diff
@@ -0,0 +1,167 @@
+                                .put(INDEX_NUMBER_OF_REPLICAS_SETTING.getKey(), numReplicas)
+                                .build();
+        final int numNodes = randomIntBetween(1, numReplicas - 1);
+        internalCluster().ensureAtMostNumDataNodes(numNodes);
```
instead of using `ensureAtMostNumDataNodes` you can also set INDEX_NUMBER_OF_REPLICAS_SETTING to `internalCluster().numDataNodes() + 1`. Simple and no need to resize the cluster during the test.

ywelsch commented on core/src/test/java/org/elasticsearch/action/admin/indices/create/CreateIndexWaitsIT.java:
```diff
@@ -0,0 +1,167 @@
+        assertThrows(prepareCreate(indexName)
+                         .setSettings(settings)
+                         .setWaitForActiveShards(ActiveShardCount.from(numNodes + 1))
+                         .setTimeout("2s"), ElasticsearchTimeoutException.class);
```
having the tests wait 2 seconds is way too long. Please increase the timeouts to 1minute (or leave default) when we expect NOT to have timeouts, otherwise set timeout to 0 so that we instantly run into timeout.

ywelsch commented on core/src/test/java/org/elasticsearch/action/admin/indices/create/CreateIndexWaitsIT.java:
```diff
@@ -0,0 +1,167 @@
+                                .put(INDEX_NUMBER_OF_REPLICAS_SETTING.getKey(), numReplicas)
+                                .build();
+        final int numNodes = randomIntBetween(1, numReplicas + 1);
+        internalCluster().ensureAtLeastNumDataNodes(numNodes);
```
same here, we can get rid of this.

ywelsch commented on core/src/test/java/org/elasticsearch/action/admin/indices/create/CreateIndexWaitsIT.java:
```diff
@@ -0,0 +1,167 @@
+    public void testInvalidWaitsValues() {
+        final String indexName = "test-idx";
+        try {
+            prepareCreate(indexName).setWaitForActiveShards(ActiveShardCount.from(-3));
```
this actually only tests `ActiveShardCount.from(-3)` and should be a unit test instead.

ywelsch commented on core/src/test/java/org/elasticsearch/indices/cluster/ClusterStateChanges.java:
```diff
@@ -119,7 +122,7 @@ public ClusterStateChanges() {
         DestructiveOperations destructiveOperations = new DestructiveOperations(settings, clusterSettings);
         Environment environment = new Environment(settings);
-        ThreadPool threadPool = null; // it's not used
+        threadPool = new TestThreadPool("ClusterStateChanges");
```
I would prefer that we don't use a threadpool in this class, it's supposed to be side-effect free. I prefer to set request.waitForActiveShards() to NONE and make a shortcut in MetaDataCreateIndexService for NONE.

ywelsch commented on core/src/test/java/org/elasticsearch/snapshots/DedicatedClusterSnapshotRestoreIT.java:
```diff
@@ -437,11 +440,12 @@ public void testRestoreIndexWithMissingShards() throws Exception {
+        CreateIndexResponse createIndexResponse = prepareCreate("test-idx-none", 1, Settings.builder().put("number_of_shards", 6)
                 .put("index.routing.allocation.include.tag", "nowhere")
-                .put("number_of_replicas", 0)));
+                .put("number_of_replicas", 0)).setWaitForActiveShards(ActiveShardCount.NONE).setTimeout("2s").get();
```
no need to set timeout, the default is good enough

ywelsch commented on rest-api-spec/src/main/resources/rest-api-spec/api/cluster.allocation_explain.json:
```diff
@@ -10,6 +10,10 @@
           "type": "boolean",
           "description": "Return 'YES' decisions in explanation (default: false)"
+        },
+        "wait_for_active_shards": {
```
why does the explain API have this parameter?

ywelsch replied:
same question for reindex and update_by_query.

nik9000 replied:
If bulk has it then reindex, update-by-query, and delete-by-query should have it. They copied `consistency` to their bulk requests so they should copy this instead.

abeyad replied:
Its unnecessary for allocation explain API, the rest of the cases require it.

ywelsch commented on core/src/main/java/org/elasticsearch/action/admin/indices/create/CreateIndexResponse.java:
```diff
@@ -30,22 +31,43 @@
+    /**
+     * Adds custom fields to the AcknowledgedResponse's x-content.
+     */
+    public void addCustomFields(XContentBuilder builder, CreateIndexResponse response) throws IOException {
```
why do you pass the response to this method? `this` already has all information.

ywelsch commented on core/src/main/java/org/elasticsearch/rest/action/admin/indices/create/RestCreateIndexAction.java:
```diff
@@ -52,6 +56,12 @@ public void handleRequest(final RestRequest request, final RestChannel channel,
+        createIndexRequest.waitForActiveShards(ActiveShardCount.parseString(request.param("wait_for_active_shards")));
+        client.admin().indices().create(createIndexRequest, new AcknowledgedRestListener<CreateIndexResponse>(channel) {
+            @Override
+            public void addCustomFields(XContentBuilder builder, CreateIndexResponse response) throws IOException {
```
to make this simpler, I think we should add a method `addCustomFields` to the `AcknowledgedResponse` object instead and just call `response.addCustomFields(builder)` in `AcknowledgedRestListener`.

As a follow-up we can then make AcknowledgedResponse implement `StatusToXContent`.

abeyad replied:
This is not as easily achieved because for other cases, adding the custom fields requires information in the request object as well as the context of the rest request, so leaving it as is for now as discussed with @ywelsch

ywelsch commented on core/src/main/java/org/elasticsearch/action/admin/indices/create/CreateIndexResponse.java:
```diff
@@ -30,22 +31,43 @@
+
+    /**
+     * Returns true if the request timed out waiting for the required number
+     * of active shards to be started.  If the index was not successfully
```
active shards cannot be started. They are started by definition of being active ;-)

ywelsch commented on core/src/main/java/org/elasticsearch/action/admin/indices/shrink/TransportShrinkAction.java:
```diff
@@ -93,22 +94,21 @@ public void onResponse(IndicesStatsResponse indicesStatsResponse) {
-                        listener.onFailure(t);
-                    }
-                });
+                final CreateIndexRequest createIndexRequest = shrinkRequest.getShrinkIndexRequest();
```
remove this line and use `updateRequest.index()` instead below.

ywelsch commented on core/src/main/java/org/elasticsearch/action/admin/indices/shrink/TransportShrinkAction.java:
```diff
@@ -93,22 +94,21 @@ public void onResponse(IndicesStatsResponse indicesStatsResponse) {
-                });
+                final CreateIndexRequest createIndexRequest = shrinkRequest.getShrinkIndexRequest();
+                createIndexService.createIndex(updateRequest,
+                    shardsWaiter.wrapUpdateListenerWithWaiting(
```
this whole block here looks pretty much the same in all invocations. Can we make this even simpler? Maybe create a method `createIndexAndWaitForActiveShards` in `MetaDataCreateIndexService`. I've implemented it here: https://github.com/ywelsch/elasticsearch/commit/6e67ecabbfa5cc2568c0c987401e3ea521c7a330

ywelsch commented on core/src/main/java/org/elasticsearch/action/support/ActiveShardCount.java:
```diff
@@ -0,0 +1,221 @@
+     * Get an ActiveShardCount instance for the given value.  The value is first validated to ensure
+     * it is a valid shard count and throws an IllegalArgumentException if validation fails.  Valid
+     * values are any non-negative number.  Directly use {@link ActiveShardCount#DEFAULT} for the
+     * default value (which is half the shards) or {@link ActiveShardCount#ALL} to specify all the shards.
```
half the shards? please update

ywelsch replied:
ping?

ywelsch commented on core/src/main/java/org/elasticsearch/action/support/ActiveShardCount.java:
```diff
@@ -0,0 +1,221 @@
+
+    /**
+     * Resolve this instance to an actual integer value for the number of active shard counts.
+     * If {@link ActiveShardCount#ALL} are specified, then the given {@link IndexMetaData} is
```
s/are/is/

ywelsch commented on core/src/main/java/org/elasticsearch/action/support/ActiveShardCount.java:
```diff
@@ -0,0 +1,221 @@
+    public boolean enoughShardsActive(final ClusterState clusterState, final String indexName, final Settings settings) {
+        Objects.requireNonNull(clusterState);
+        Objects.requireNonNull(indexName);
+        Objects.requireNonNull(settings);
```
let's assume that if the method parameters are not marked as @Nullable that they are non-null. Otherwise we clutter the codebase everywhere with these checks.

ywelsch commented on core/src/main/java/org/elasticsearch/action/support/ActiveShardsWaiter.java:
```diff
@@ -0,0 +1,145 @@
+                            @Override
+                            public boolean apply(ClusterState previousState, ClusterState.ClusterStateStatus previousStatus,
+                                                 ClusterState newState, ClusterState.ClusterStateStatus newStatus) {
+                                return waitForActiveShards.enoughShardsActive(newState, indexName, settings);
```
here (and two lines below) you pass the node settings and not the index settings. How did this pass any integration tests? I would also prefer to change the settings parameter of this method to indexmetadata to avoid these confusions.

ywelsch commented on core/src/main/java/org/elasticsearch/action/support/ActiveShardsWaiter.java:
```diff
@@ -0,0 +1,145 @@
+            @Override
+            public void onFailure(Throwable t) {
+                if (t instanceof IndexAlreadyExistsException) {
+                    logger.trace("[{}] failed to create", t, request.index());
```
why is the logging for "failed to create index" in this class here?

ywelsch commented on core/src/test/java/org/elasticsearch/action/support/ActiveShardCountTests.java:
```diff
@@ -0,0 +1,294 @@
+        assertTrue(waitForActiveShards.enoughShardsActive(clusterState, indexName, Settings.EMPTY));
+    }
+
+    private void runTestForOneActiveShard(final ActiveShardCount activeShardCount) {
```
dead code?

ywelsch commented on rest-api-spec/src/main/resources/rest-api-spec/api/indices.create.json:
```diff
@@ -13,6 +13,10 @@
         }
       },
       "params": {
+        "wait_for_active_shards": {
```
what about the other rest actions that support this parameter now? e.g. shrink.

Can you also add a rest test?

abeyad commented:
@ywelsch I added your `MetaDataCreateIndexService` refactoring commit and added this commit to address other code review comments: https://github.com/elastic/elasticsearch/pull/18985/commits/237dc92c6dc3e1a4b25917c912e3e373b4cfbdf6

ywelsch commented on core/src/main/java/org/elasticsearch/action/support/ActiveShardCount.java:
```diff
@@ -140,17 +137,16 @@ public static ActiveShardCount parseString(final String str) {
             return true;
         }
+        final IndexMetaData indexMetaData = clusterState.metaData().index(indexName);
+        assert indexMetaData != null;
```
what if the index was deleted while we wait on shards?

abeyad replied:
Its a bit of a weird situation, as it would mean the cluster state was independently monitored to know the index exists, and then a delete index request was issued while still waiting for shards after the create request.  We should definitely still account for it, but should the create index response indicate that this occurred, instead of just returning `{ acknowledged: true, timed_out_waiting_for_shards: false }`?

ywelsch replied:
I'm fine with `{ acknowledged: true, timed_out_waiting_for_shards: false }`. Throwing an exception might just be annoying in this case?

abeyad replied:
Ok, so we will treat it as if all required shards have become active and we are done waiting.

abeyad replied:
I will add a test for this case too

ywelsch commented on rest-api-spec/src/main/resources/rest-api-spec/api/indices.shrink.json:
```diff
@@ -25,6 +25,10 @@
+        },
+        "wait_for_active_shards": {
+          "type" : "string",
+          "description" : "Set the number of active shards to wait for before proceeding with the operation."
```
we can be more specific here: which active shards? of the index that we shrink from or to?

abeyad replied:
the index being shrunk to, I will amend the description here and above for rollover

ywelsch commented on rest-api-spec/src/main/resources/rest-api-spec/api/indices.rollover.json:
```diff
@@ -25,6 +25,10 @@
+        },
+        "wait_for_active_shards": {
+          "type" : "string",
+          "description" : "Set the number of active shards to wait for before proceeding with the operation."
```
same here, more specific which shards we are talking about here.

ywelsch commented on rest-api-spec/src/main/resources/rest-api-spec/test/indices.create/10_basic.yaml:
```diff
@@ -31,6 +31,20 @@
   - match: { test_index.settings.index.number_of_replicas: "0"}
 
 ---
+"Create index with too large wait_for_active_shards":
```
let's also add a happy test, for e.g. with "all"

abeyad commented on core/src/main/java/org/elasticsearch/action/support/ActiveShardsWaiter.java:
```diff
@@ -0,0 +1,140 @@
+                            @Override
+                            public boolean apply(ClusterState previousState, ClusterState.ClusterStateStatus previousStatus,
+                                                 ClusterState newState, ClusterState.ClusterStateStatus newStatus) {
+                                return waitForActiveShards.enoughShardsActive(newState, indexName);
```
i will add a test to put “index.wait_for_active_shards” in the index settings with a value different than the default, then ensure it used that value found in the settings

ywelsch commented on core/src/main/java/org/elasticsearch/action/support/ActiveShardsWaiter.java:
```diff
@@ -0,0 +1,140 @@
+ * This class provides primitives for waiting for a configured number of shards
+ * to become active before sending a response on an {@link ActionListener}.
+ */
+public class ActiveShardsWaiter extends AbstractComponent {
```
Maybe call this ActiveShardsObserver (similar to ClusterStateObserver)?

abeyad replied:
Done

ywelsch commented on core/src/main/java/org/elasticsearch/index/IndexSettings.java:
```diff
@@ -127,6 +128,17 @@
 
+    /**
+     * The number of shard copies required to be active before proceed with the
+     * write operation or returning from an index creation operation.
```
this only applies to index creation operation, not to write operations (especially with the name "index." in the setting...

abeyad replied:
Done, sorry, remnants from the previous PR when the setting was used for both.

ywelsch commented on core/src/test/java/org/elasticsearch/action/admin/indices/create/CreateIndexWaitsIT.java:
```diff
@@ -0,0 +1,129 @@
+ * Tests that the index creation operation waits for the appropriate
+ * number of active shards to be started before returning.
+ */
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST)
```
why is the scope == TEST?

abeyad replied:
Remnant from when I was trying to use numDataNodes to manipulate the cluster with this test, I've removed it.

ywelsch commented on core/src/test/java/org/elasticsearch/action/admin/indices/create/CreateIndexWaitsIT.java:
```diff
@@ -0,0 +1,129 @@
+
+    public void testCreateIndexNotEnoughActiveShardsTimesOut() throws Exception {
+        final String indexName = "test-idx";
+        final int numReplicas = internalCluster().numDataNodes() + randomIntBetween(2, 4);
```
why not `randomIntBetween(0, 4);` which is `randomInt(4)`? We just need to make sure not all copies get a node.

abeyad replied:
no good reason, I've changed it

ywelsch commented on core/src/test/java/org/elasticsearch/action/admin/indices/create/CreateIndexWaitsIT.java:
```diff
@@ -0,0 +1,129 @@
+                                .build();
+        assertTrue(prepareCreate(indexName)
+                       .setSettings(settings)
+                       .setWaitForActiveShards(ActiveShardCount.from(internalCluster().numDataNodes() + 1))
```
instead + 1 it can be + randomIntBetween(1,4)

abeyad replied:
The reason its just +1 is because if I did randomIntBetween(1,4), I could get a random number greater than the numReplicas + 1, in which case I will get an IllegalArgumentException instead of just timing out waiting for shards  (because I would've tried specifying a wait_for_active_shards > possible number of shard copies, so it would be impossible to achieve) i

ywelsch replied:
randomIntBetween(1, numReplicas + 1)?

ywelsch commented on core/src/test/java/org/elasticsearch/action/admin/indices/create/CreateIndexWaitsIT.java:
```diff
@@ -0,0 +1,129 @@
+                                .put(INDEX_NUMBER_OF_SHARDS_SETTING.getKey(), randomIntBetween(2, 7))
+                                .put(INDEX_NUMBER_OF_REPLICAS_SETTING.getKey(), internalCluster().numDataNodes() + randomIntBetween(0, 3))
+                                .build();
+        ActiveShardCount waitForActiveShards = ActiveShardCount.from(randomIntBetween(0, internalCluster().numDataNodes() - 1));
```
why - 1?

abeyad replied:
no good reason, changed

ywelsch commented on core/src/test/java/org/elasticsearch/action/admin/indices/create/CreateIndexWaitsIT.java:
```diff
@@ -0,0 +1,129 @@
+        final String indexName = "test-idx";
+        Settings settings = Settings.builder()
+                                .put(indexSettings())
+                                .put(INDEX_NUMBER_OF_SHARDS_SETTING.getKey(), randomIntBetween(2, 7))
```
any reason not to run this test for 1 shard?

abeyad replied:
no good reason, I've changed it

abeyad commented:
@ywelsch https://github.com/elastic/elasticsearch/pull/18985/commits/3645e6a650b2b1632d359db7711a09f7cdc76975 addresses the latest comments

ywelsch commented on core/src/test/java/org/elasticsearch/action/admin/indices/create/CreateIndexWaitsIT.java:
```diff
@@ -19,26 +19,28 @@
  * number of active shards to be started before returning.
  */
-@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST)
 public class CreateIndexWaitsIT extends ESIntegTestCase {
```
This could now be called ActiveShardsObserverIT

ywelsch commented on core/src/test/java/org/elasticsearch/action/admin/indices/create/CreateIndexWaitsIT.java:
```diff
@@ -120,10 +122,66 @@ public void testCreateIndexWaitsForAllActiveShards() throws Exception {
+                       .setTimeout("100ms")
+                       .get()
+                       .isTimedOutWaitingForShards());
+    }
```
how about extending this test in the following way:

create index with WAIT_FOR_ACTIVE_SHARDS_SETTING higher than number of data nodes and timeout 1h.
check that future is not resolved
update index settings WAIT_FOR_ACTIVE_SHARDS_SETTING to number lower than data nodes.
check that previous future is now resolved (future.get(10, seconds)).

ywelsch commented on core/src/main/java/org/elasticsearch/action/admin/indices/rollover/TransportRolloverAction.java:
```diff
@@ -111,42 +110,45 @@ public void onResponse(IndicesStatsResponse statsResponse) {
+                                    new ActionListener<ClusterStateUpdateResponse>() {
+                                        @Override
+                                        public void onResponse(ClusterStateUpdateResponse clusterStateUpdateResponse) {
+                                            // TODO: what happens if this ClusterStateUpdateResponse is not acknowledged?
```
I added this TODO because I think that this needs to be clarified. Either we say that we ignore acknowledgment of the aliases update and document that here instead of the TODO or we chose to only do the wait on active shards if it gets acknowledged. I think that I'm more in favour of the former...

abeyad replied:
I also am in favor of the former, I didn't change anything here originally because that is how the rollover action worked.  If it should be changed to do something else on `acknowledged = false`, we can make that in a separate PR?

ywelsch commented on core/src/main/java/org/elasticsearch/action/delete/TransportDeleteAction.java:
```diff
@@ -70,7 +70,10 @@ public TransportDeleteAction(Settings settings, TransportService transportServic
         ClusterState state = clusterService.state();
         if (autoCreateIndex.shouldAutoCreate(request.index(), state)) {
-            createIndexAction.execute(task, new CreateIndexRequest().index(request.index()).cause("auto(delete api)").masterNodeTimeout(request.timeout()), new ActionListener<CreateIndexResponse>() {
+            CreateIndexRequest createIndexRequest = new CreateIndexRequest()
```
nothing changed in this class, please revert

ywelsch commented on core/src/main/java/org/elasticsearch/action/support/ActiveShardCount.java:
```diff
@@ -0,0 +1,224 @@
+     * Returns true iff the active shard count in the shard routing table is enough
+     * to meet the required shard count represented by this instance.
+     */
+    public boolean checkWaitForActiveShards(final IndexShardRoutingTable shardRoutingTable, final IndexMetaData indexMetaData) {
```
I would give this method the same name as the method above: `enoughShardsActive`. 
One operates at index, the other at the shard level. To make things even more explicit, the above method could have parameters
(IndexRoutingTable, IndexMetaData) and this one (IndexShardRoutingTable, IndexMetaData). This would also leave it up to the caller to deal with `indexMetaData == null`.

abeyad replied:
I like the idea of changing the method name to `enoughShardsActive`, but I don't think it makes things simpler to have the (IndexRoutingTable, IndexMetaData) method signature, because it requires the indexMetaData null check etc to be done at the caller, and there are two places calling it (so far), so it doesn't make things simpler in my view.  If you still want it though, I can make the change.

ywelsch commented on core/src/main/java/org/elasticsearch/action/support/ActiveShardsObserver.java:
```diff
@@ -0,0 +1,136 @@
+     * @return ActionListener for responding to index creation cluster state events and sending the action response
+     *         over the main listener, after waiting for the requested number of shards to be active
+     */
+    public <T> ActionListener<ClusterStateUpdateResponse> wrapUpdateListenerWithWaiting(final CreateIndexClusterStateUpdateRequest request,
```
maybe just call this waitForActiveShards?

ywelsch commented on core/src/main/java/org/elasticsearch/action/support/ActiveShardsObserver.java:
```diff
@@ -0,0 +1,136 @@
+                        observer.waitForNextChange(observerListener, shardsAllocatedPredicate, timeout);
+
+                    } catch (Exception ex) {
+                        logger.debug("[{}] index creation failed on waiting for shards", request.index());
```
what kind of exceptions are we catching here?

abeyad replied:
Upon investigation, the only exception that can be thrown is the ElasticsearchException if the cluster state is already being observed by the observer.  This observer is only used within the confines of this method, and is only called here once, so it should not be possible for this exception to be thrown.  However, I feel this try catch block is relevant anyway because if an exception happens to be thrown by the call to waitForNextChange (or some logic is added to the ClusterStateObserver that enables an exception to be thrown), then at least we wouldn't block on the failure here and would send it immediately to the client.

ywelsch commented on core/src/main/java/org/elasticsearch/action/update/TransportUpdateAction.java:
```diff
@@ -118,7 +118,10 @@ public static void resolveAndValidateRouting(MetaData metaData, String concreteI
         // if we don't have a master, we don't have metadata, that's fine, let it find a master using create index API
         if (autoCreateIndex.shouldAutoCreate(request.index(), clusterService.state())) {
-            createIndexAction.execute(new CreateIndexRequest().index(request.index()).cause("auto(update api)").masterNodeTimeout(request.timeout()), new ActionListener<CreateIndexResponse>() {
+            CreateIndexRequest createIndexRequest = new CreateIndexRequest()
```
same as above, revert this class (no change was made).

ywelsch commented on core/src/main/java/org/elasticsearch/index/IndexSettings.java:
```diff
@@ -127,6 +128,16 @@
+     */
+    public static final Setting<ActiveShardCount> WAIT_FOR_ACTIVE_SHARDS_SETTING =
+        new Setting<>("index.wait_for_active_shards",
+                         (String) null,
```
we could also add the default here: "1". This would simplify the "resolve" code a bit.

abeyad replied:
I changed it to "1", though I don't think it would simplify the `resolve` code nor the `parseString` code.  I can remove the null check in `parseString` if I also pass in defaults to all the rest actions `request.param("wait_for_active_shards")`, though I'm not sure what that buys us?

ywelsch commented:
Left a few more comments (it's getting close...)

abeyad commented:
@ywelsch https://github.com/elastic/elasticsearch/pull/18985/commits/b0d4f2865cd64f789e5539c869a78286f680d0b5 addresses the latest comments

bleskes commented on core/src/main/java/org/elasticsearch/action/support/ActiveShardCount.java:
```diff
@@ -0,0 +1,224 @@
+
+    private final int value;
+
+    private ActiveShardCount(final int value) {
```
does this really need to be an instance? can't we use a `int` and have utility static methods hard?

ywelsch replied:
I prefer to encapsulate this in a class instead of floating magic values around (-1 and -2) making it hard to ensure they are properly used everywhere.

bleskes replied:
I get that, but given that this going to be used in the indexing logic as well (we should share the code) - I rather avoid the boxing. We can add request level/serialization validation that the values are proper.

abeyad replied:
@bleskes Its not clear to me how this would cause performance problems... with the Java API, my guess is most people will use the pre-created defaults (DEFAULT, ALL, NONE, ONE), and in the case of the REST API, it would just be one lookup per indexing request and assuming bulk requests most of the time, the cost seems to be pretty well amortized?

bleskes replied:
ok. I see the parsing code now reuses those object instances in it's parsing logic. I personally think this is overly engineered but not enough to ask you to change it if you prefer it this way. Up to you and @ywelsch then.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/ActiveShardsObserver.java:
```diff
@@ -0,0 +1,136 @@
+     * @return ActionListener for responding to index creation cluster state events and sending the action response
+     *         over the main listener, after waiting for the requested number of shards to be active
+     */
+    public <T> ActionListener<ClusterStateUpdateResponse> waitForActiveShards(final CreateIndexClusterStateUpdateRequest request,
```
does it make sense for this method to generate a listener? if we really need this, let's reflect it in the name. Reading how it's used, I think it will be cleaner to have the relevant listener created by the caller and have this method do something:

```
new ActionListener<ClusterStateUpdateResponse>() {
       public void onResponse(ClusterStateUpdateResponse response) {
                      if (response.isAcknowledged()) {
                           waitForActiveShards(request, actionListener, onResult); <--- call it here.
                     } else {
                           actionListener.onResponse(response);
                     } 
      } 

     onFailure(..) { 
          ....
     } 

```

Also - double checking - do we need both actionListener and onResult ? it seems like a double callback structure. Maybe it's just a naming issue, but it seems we can keep this simple and just let this method tell use if it timed out or not? (given my suggestion to to create the wrapper listener on the outside?

bleskes commented on core/src/main/java/org/elasticsearch/action/support/ActiveShardsObserver.java:
```diff
@@ -0,0 +1,136 @@
+            @Override
+            public void onResponse(ClusterStateUpdateResponse response) {
+                if (response.isAcknowledged()) {
+                    final String indexName = request.index();
```
make these parmaters to this method? then it doesn't really need to know about index creation requests.

abeyad replied:
That's how it was before, it was changed to enable less params to be passed into the method.  I agree with your suggestion though, so I'll change it back.

bleskes commented on core/src/main/java/org/elasticsearch/action/support/ActiveShardsObserver.java:
```diff
@@ -0,0 +1,136 @@
+                        // wait for the configured number of active shards to be allocated before returning
+                        final ClusterStateObserver observer =
+                            new ClusterStateObserver(clusterService, logger, threadPool.getThreadContext());
+                        final ClusterStateObserver.ChangePredicate shardsAllocatedPredicate = new ClusterStateObserver.ChangePredicate() {
```
see ValidationPredicate

bleskes commented on core/src/main/java/org/elasticsearch/index/IndexSettings.java:
```diff
@@ -127,6 +128,16 @@
+    /**
+     * The number of shard copies required to be active before returning from an index creation operation.
+     */
+    public static final Setting<ActiveShardCount> WAIT_FOR_ACTIVE_SHARDS_SETTING =
```
how does this play with the parameter of the index creation request? does it serve as a default to the explicit request parameter? do we really need the request parameter if so? (i.e., only use the setting)

abeyad replied:
@bleskes This parameter serves as a default if wait_for_active_shards is not specified in the request.  We could just use the setting and it would simplify things.  I think originally I was trying to keep consistent with the way indexing requests enable specifying the write consistency param, but upon reflection, it does seem unnecessary here.  Would you like me to remove it?

bleskes replied:
well, taking this a step further, the setting idea was to be used as a replacement for the current action setting that is a default for the indexing (i.e. replication) operations. I think that's a good idea as well as making the setting an index setting. I was thinking that if we have that setting, do we really need a request level parameter for the index creation or can we just fall back to the setting (especially if they are dynamic). Thinking about it more, I think request parameters OK. I'm not sure about `index. wait_for_active_shards` . Can't come up with a better one though. Will think about it more..

abeyad commented:
@bleskes @ywelsch https://github.com/elastic/elasticsearch/pull/18985/commits/ad4554645e66453b68dc1e377e772414597f05fa uses Boaz's patch

https://github.com/elastic/elasticsearch/pull/18985/commits/bc6b0456d3a5b07641f13ed12f1df5e0065bc56f addresses the other code review comments from Boaz

bleskes commented on core/src/main/java/org/elasticsearch/action/support/ActiveShardCount.java:
```diff
@@ -0,0 +1,209 @@
+
+    public static final ActiveShardCount DEFAULT = new ActiveShardCount(ACTIVE_SHARD_COUNT_DEFAULT);
+    public static final ActiveShardCount ALL = new ActiveShardCount(ALL_ACTIVE_SHARDS);
+    public static final ActiveShardCount NONE = new ActiveShardCount(0);
```
@clintongormley mentioned that NONE doesn't have many external usages (we only use it for index auto creation) so we might want to drop the special naming and use `0`. I will keep the object reuse in parsing.

bleskes replied:
and.. looking at the parsing logic this is indeed internal and we don't accept none from  strings. Sorry for the noise.

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/create/CreateIndexRequest.java:
```diff
@@ -440,6 +443,27 @@ public CreateIndexRequest updateAllTypes(boolean updateAllTypes) {
+    }
+
+    /**
+     * Sets the number of shard copies that must be active for index creation to return.
```
we should soften the language here. We can return before these are active (with a time out flag))

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/create/CreateIndexRequestBuilder.java:
```diff
@@ -249,4 +250,20 @@ public CreateIndexRequestBuilder setUpdateAllTypes(boolean updateAllTypes) {
     }
+
+    /**
+     * Sets the number of shard copies that must be active for index creation to return.
```
same comment about the language.

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/create/CreateIndexResponse.java:
```diff
@@ -30,22 +31,39 @@
+     * Returns true if the requisite number of shards were started before
+     * returning from the index creation operation.
+     */
+    public boolean isShardsAcked() {
```
can we document the relationship with isAcknowledged() ? i.e., if that is false, shardsAcked is also false

bleskes replied:
while at it, can we add an assert on the constructor?

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/create/TransportCreateIndexAction.java:
```diff
@@ -77,24 +75,12 @@ protected void masterOperation(final CreateIndexRequest request, final ClusterSt
-        });
+        createIndexService.createIndex(updateRequest, ActionListener.wrap(response ->
+            listener.onResponse(new CreateIndexResponse(response.isAcknowledged(), response.isShardsAcked())),
+            listener::onFailure));
```
we lost the logging here, did we get them somewhere else?

abeyad replied:
Its now here: https://github.com/elastic/elasticsearch/pull/18985/files/ad4554645e66453b68dc1e377e772414597f05fa#diff-f88ab41a466f9c0cd65780dba9fa5355R442

bleskes replied:
never mind, I saw them later on

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/rollover/RolloverRequest.java:
```diff
@@ -206,4 +207,19 @@ public void source(BytesReference source) {
     }
 
+    /**
+     * Sets the number of shard copies that must be active for index creation to return.
```
same language comment

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/rollover/RolloverResponse.java:
```diff
@@ -95,6 +98,14 @@ public boolean isRolledOver() {
+     * Returns true if the requisite number of shards were started before
+     * returning from the index creation operation.
+     */
+    public boolean isShardsAcked() {
```
can also expose the index level ack, or neither? this is confusing imo.

abeyad replied:
@bleskes the `RolloverResponse` seems to be missing fields because `rolledOver` only seems to indicate if `dryRun=false` and the conditions were met, but doesn't give any indication of the success of either the create index cluster state update operation nor the alias switch cluster state update operation.  I've added both in this commit: https://github.com/elastic/elasticsearch/pull/18985/commits/6c2e48a78369318330e5357d5c84d885c0a9423b

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/rollover/RolloverResponse.java:
```diff
@@ -95,6 +98,14 @@ public boolean isRolledOver() {
     }
 
+    /**
+     * Returns true if the requisite number of shards were started before
```
we should clearly state that these are the shard of the new index

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/shrink/ShrinkRequest.java:
```diff
@@ -126,6 +126,21 @@ public String getSourceIndex() {
     }
 
+    /**
+     * Sets the number of shard copies that must be active for index creation to return.
```
same comment about language

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/shrink/TransportShrinkAction.java:
```diff
@@ -93,22 +91,8 @@ public void onResponse(IndicesStatsResponse indicesStatsResponse) {
-                    }
-                });
+                createIndexService.createIndex(updateRequest, ActionListener.wrap(response ->
+                    listener.onResponse(new ShrinkResponse(response.isAcknowledged(), response.isShardsAcked())), listener::onFailure));
```
lost the logging here too?

bleskes replied:
never mind, I saw them later on

bleskes commented on core/src/main/java/org/elasticsearch/cluster/metadata/MetaDataCreateIndexService.java:
```diff
@@ -176,7 +182,24 @@ public void validateIndexName(String index, ClusterState state) {
     }
 
-    public void createIndex(final CreateIndexClusterStateUpdateRequest request, final ActionListener<ClusterStateUpdateResponse> listener) {
+    public void createIndex(final CreateIndexClusterStateUpdateRequest request,
```
can we add java docs about the behavior and relationship between index level ack and shard level ack?

bleskes commented on core/src/test/java/org/elasticsearch/action/admin/cluster/allocation/ClusterAllocationExplainIT.java:
```diff
@@ -57,8 +57,8 @@ public void testDelayShards() throws Exception {
-                .put(IndexMetaData.INDEX_NUMBER_OF_REPLICAS_SETTING.getKey(), 1)).get();
-        ensureGreen("test");
+                .put(IndexMetaData.INDEX_NUMBER_OF_REPLICAS_SETTING.getKey(), 1))
+                .setWaitForActiveShards(ActiveShardCount.ALL).get();
```
ensure green has the nice reporting if we fail to get to green. Can we have an equivalent utility method that takes a create index response and  check for shard acked, and log things? maybe a variant of `assertAcked`

bleskes commented:
This looks great. I left some very minor comments. I do think we miss the change to TransportIndexAction and TransportBulkAction where we set the waitForActive shards to 0. 

Of course, @ywelsch should also LGTM this as he is the main reviewer

abeyad commented:
@bleskes Thank you for the review.  https://github.com/elastic/elasticsearch/pull/18985/commits/6c2e48a78369318330e5357d5c84d885c0a9423b addresses your code review comments.

bleskes commented on core/src/main/java/org/elasticsearch/action/admin/indices/rollover/RolloverResponse.java:
```diff
@@ -39,25 +39,31 @@
     private static final String DRY_RUN = "dry_run";
     private static final String ROLLED_OVER = "rolled_over";
     private static final String CONDITIONS = "conditions";
+    private static final String CREATE_INDEX_ACKNOWLEDGED = "create_index_acknowledged";
```
hmm.. i'm not sure we want to expose the complexity of the internal two steps. For example if the alias is acked and the index isn't, it doesn't really mean that the index isn't known by the nodes - having processed the alias change guarantees it.  I think we can keep a simple acknowledge like in other responses, which means all nodes know about the new index and the new alias? it maybe then that with rollover the shards are acked but not the request. If we want to avoid that , I will only wait for shards active after the alias change is done.

abeyad replied:
> it maybe then that with rollover the shards are acked but not the request. If we want to avoid that , I will only wait for shards active after the alias change is done.

That's kind of why its a bit nebulous.  The semantics of `acknowledged` on `ClusterStateUpdateResponse` should be the same, but the trick here is that we have two different cluster state updates - one for index creation and one for the alias.  We can move the shard waiting to after the alias swap, but then we also need to change the `MetaDataCreateIndexService` to make `onlyCreateIndex` public, and probably change the names.... `createIndex` only creates the index, `createIndexAndWait` includes the waiting... if they are to be public methods.. thoughts?

bleskes replied:
we can explicitly set the the waitForShards of the index creation to NONE and use the setting of the request to wait after the alias was acked. All variants of this are not ideal, so I would go for what you find the simplest - I tend to agree that consistency is the best here and thus leaning towards what I just described.

bleskes commented on test/framework/src/main/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertions.java:
```diff
@@ -739,4 +740,14 @@ public static void assertDirectoryExists(Path dir) {
+     * Assert that an index creation was fully acknowledged, meaning that both the index creation cluster
+     * state update was successful and that the requisite number of shard copies were started before returning.
+     */
+    public static void assertCreateIndexAcked(CreateIndexResponse response) {
```
just call this assertAcked? the parameter is enough to distinguish what it does?

abeyad commented:
@bleskes https://github.com/elastic/elasticsearch/pull/18985/commits/fcf60eeff9d69365c1c0740573d4bd5bf6734a9c addresses the latest comments

ywelsch commented on core/src/main/java/org/elasticsearch/action/support/ActiveShardCount.java:
```diff
@@ -0,0 +1,209 @@
+            case ALL_ACTIVE_SHARDS:
+                return ALL;
+            case 0:
+                return NONE;
```
we "special case" NONE here but not ONE, maybe it's simpler just to remove this method as well as the `validateValue` one and use `new ActiveShardCount(...)` in the two places it's currently used (and also ad
```
if (value < -2) {
  throw new IllegalArgumentException(...)
}
```
to the constructor.

ywelsch replied:
I you decide to go this route you should also remember to replace the reference equality checks (`this == ActiveShardCount.NONE`) by equals checks or by looking at value (`this.value == 0`).

abeyad replied:
The reason I preferred the current way is that we don't have to recreate new objects for what will most likely be the same values re-used over and over again.

ywelsch replied:
ok sounds good. maybe add the `case 1:` then.

ywelsch commented on core/src/main/java/org/elasticsearch/action/support/ActiveShardCount.java:
```diff
@@ -0,0 +1,209 @@
+            try {
+                val = Integer.parseInt(str);
+            } catch (NumberFormatException e) {
+                throw new IllegalArgumentException("cannot parse ActiveShardCount[" + str + "]");
```
can you also add the original exception as a parameter?

ywelsch commented on core/src/main/java/org/elasticsearch/action/support/ActiveShardsObserver.java:
```diff
@@ -0,0 +1,105 @@
+                                    final ActiveShardCount activeShardCount,
+                                    final TimeValue timeout,
+                                    final Consumer<Boolean> onResult,
+                                    final Consumer<Exception> onError) {
```
maybe call this `onFailure`

ywelsch commented on core/src/main/java/org/elasticsearch/cluster/ack/CreateIndexClusterStateUpdateResponse.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    /**
+     * Returns whether the requisite number of shards started before the completion of the operation or timing out.
```
remove "or timing out".

ywelsch commented on core/src/test/java/org/elasticsearch/snapshots/DedicatedClusterSnapshotRestoreIT.java:
```diff
@@ -441,11 +444,12 @@ public void testRestoreIndexWithMissingShards() throws Exception {
                 .put("index.routing.allocation.include.tag", "nowhere")
-                .put("number_of_replicas", 0)));
+                .put("number_of_replicas", 0)).setWaitForActiveShards(ActiveShardCount.NONE).get();
+        assertTrue(createIndexResponse.isAcknowledged());
```
assertAcked?

ywelsch commented on core/src/test/java/org/elasticsearch/action/support/ActiveShardsObserverIT.java:
```diff
@@ -0,0 +1,157 @@
+        logger.info("--> ensure the create index request completes");
+        CreateIndexResponse response = responseListener.get();
+        assertAcked(response);
+        assertTrue(response.isShardsAcked());
```
This is covered by assertAcked? (same for other occurrences in this class)

ywelsch commented on core/src/test/java/org/elasticsearch/action/support/ActiveShardsObserverIT.java:
```diff
@@ -0,0 +1,157 @@
+                       .setWaitForActiveShards(ActiveShardCount.from(randomIntBetween(numDataNodes + 1, numReplicas + 1)))
+                       .setTimeout("100ms")
+                       .get()
+                       .isShardsAcked());
```
we should check here that acked == true but shardAcked == false

abeyad replied:
with the low timeout value, its not strictly guaranteed that the index creation is completed in time, that's why i didn't add that assertion

ywelsch commented on core/src/test/java/org/elasticsearch/action/support/ActiveShardsObserverIT.java:
```diff
@@ -0,0 +1,157 @@
+                       .setWaitForActiveShards(randomBoolean() ? ActiveShardCount.from(1) : ActiveShardCount.ALL)
+                       .setTimeout("100ms")
+                       .get()
+                       .isShardsAcked());
```
same as my comment below

ywelsch commented on rest-api-spec/src/main/resources/rest-api-spec/test/indices.create/10_basic.yaml:
```diff
@@ -31,6 +31,35 @@
+          settings:
+            number_of_replicas: 5
+
+  - match: { shards_acknowledged: false }
```
also  `match: { acknowledged: true }` ?

abeyad replied:
same reason, the low timeout doesn't necessarily guarantee index creation completed in time, and i didn't want the tests failing because of that

ywelsch commented:
I left minor comments. Once addressed, feel free to push. LGTM! Thanks @abeyad.

abeyad commented:
@ywelsch @bleskes Thank you very much for all the valuable feedback!
